From e0404d7898f89abfce72bcf910f15c812f019221 Mon Sep 17 00:00:00 2001
From: q66 <q66@chimera-linux.org>
Date: Sun, 4 Aug 2024 00:42:35 +0200
Subject: [PATCH] plumb in support for externally provided allocator libraries

The external static library can provide the following symbols:

* __malloc_init
* __malloc_atfork
* __malloc_donate
* __malloc_tls_teardown
* __malloc_tls_default
* __libc_calloc
* __libc_free
* __libc_malloc_impl
* __libc_realloc
* aligned_alloc
* malloc_usable_size

It can access the following variables via extern:

* __aligned_alloc_replaced
* __malloc_replaced

Patch-Source: https://github.com/chimera-linux/cports/blob/master/main/musl/patches/0001-plumb-in-support-for-externally-provided-allocator-l.patch
---
 Makefile                     | 17 +++++++++++++----
 ldso/dynlink.c               |  9 ++++++++-
 mimalloc/readme.md           |  1 +
 src/env/__init_tls.c         |  8 +++++++-
 src/exit/exit.c              |  2 ++
 src/internal/pthread_impl.h  | 13 +++++++++++++
 src/malloc/calloc.c          |  4 ++++
 src/malloc/external/empty.h  |  1 +
 src/malloc/libc_calloc.c     |  4 ++++
 src/malloc/mallocng/malloc.c |  2 ++
 src/thread/pthread_create.c  |  7 +++++++
 11 files changed, 62 insertions(+), 6 deletions(-)
 create mode 100644 src/malloc/external/empty.h

diff --git a/Makefile b/Makefile
index e04d1f58..28486642 100644
--- a/Makefile
+++ b/Makefile
@@ -23,6 +23,7 @@ BASE_GLOBS = $(addsuffix /*.c,$(SRC_DIRS))
 ARCH_GLOBS = $(addsuffix /$(ARCH)/*.[csS],$(SRC_DIRS))
 BASE_SRCS = $(sort $(wildcard $(BASE_GLOBS)))
 ARCH_SRCS = $(sort $(wildcard $(ARCH_GLOBS)))
+MALLOC_OBJS = $(sort $(wildcard $(srcdir)/src/malloc/$(MALLOC_DIR)/*.o))
 BASE_OBJS = $(patsubst $(srcdir)/%,%.o,$(basename $(BASE_SRCS)))
 ARCH_OBJS = $(patsubst $(srcdir)/%,%.o,$(basename $(ARCH_SRCS)))
 REPLACED_OBJS = $(sort $(subst /$(ARCH)/,/,$(ARCH_OBJS)))
@@ -32,8 +33,8 @@ LIBC_OBJS = $(filter obj/src/%,$(ALL_OBJS)) $(filter obj/compat/%,$(ALL_OBJS))
 LDSO_OBJS = $(filter obj/ldso/%,$(ALL_OBJS:%.o=%.lo))
 CRT_OBJS = $(filter obj/crt/%,$(ALL_OBJS))
 
-AOBJS = $(LIBC_OBJS)
-LOBJS = $(LIBC_OBJS:.o=.lo)
+AOBJS = $(LIBC_OBJS) $(MALLOC_OBJS)
+LOBJS = $(LIBC_OBJS:.o=.lo) $(MALLOC_OBJS)
 GENH = obj/include/bits/alltypes.h obj/include/bits/syscall.h
 GENH_INT = obj/src/internal/version.h
 IMPH = $(addprefix $(srcdir)/, src/internal/stdio_impl.h src/internal/pthread_impl.h src/internal/locale_impl.h src/internal/libc.h)
@@ -133,6 +134,11 @@ $(CRT_OBJS): CFLAGS_ALL += -DCRT -fno-lto
 
 $(LOBJS) $(LDSO_OBJS): CFLAGS_ALL += -fPIC
 
+ifneq (mallocng,$(MALLOC_DIR))
+obj/src/malloc/calloc.lo: CFLAGS_ALL += -DLIBC_CALLOC_EXTERNAL
+obj/src/malloc/libc_calloc.lo: CFLAGS_ALL += -DLIBC_CALLOC_EXTERNAL
+endif
+
 CC_CMD = $(CC) $(CFLAGS_ALL) -c -o $@ $<
 
 # Choose invocation of assembler to be used
@@ -142,6 +148,9 @@ else
 	AS_CMD = $(CC_CMD)
 endif
 
+$(EXTRA_OBJ): $(GENH) $(IMPH)
+	$(CC) -I$(srcdir)/mimalloc/include $(CFLAGS_ALL:-nostdinc=) -std=gnu11 -fPIC -O3 -DNDEBUG -fvisibility=hidden -c -o $(EXTRA_OBJ) $(srcdir)/mimalloc/src/mimalloc.c
+
 obj/%.o: $(srcdir)/%.s
 	$(AS_CMD)
 
@@ -163,11 +172,11 @@ obj/%.lo: $(srcdir)/%.c $(GENH) $(IMPH)
 unpack_glibc_string:
 	test -d glibc-string-ar || mkdir glibc-string-ar && cd glibc-string-ar && $(AR) -x ../libglibc-string.a
 
-lib/libc.so: $(LOBJS) $(LDSO_OBJS) unpack_glibc_string
+lib/libc.so: $(LOBJS) $(LDSO_OBJS) $(EXTRA_OBJ) unpack_glibc_string
 	$(CC) $(CFLAGS_ALL) $(LDFLAGS_ALL) -nostdlib -shared \
 	-Wl,-e,_dlstart -o $@ $(LOBJS) glibc-string-ar/* $(LDSO_OBJS) $(LIBCC)
 
-lib/libc.a: $(AOBJS) unpack_glibc_string
+lib/libc.a: $(AOBJS) $(EXTRA_OBJ) unpack_glibc_string
 	rm -f $@
 	$(AR) rc $@ $(AOBJS) glibc-string-ar/*
 	$(RANLIB) $@
diff --git a/ldso/dynlink.c b/ldso/dynlink.c
index 3ff540d6..00f84089 100644
--- a/ldso/dynlink.c
+++ b/ldso/dynlink.c
@@ -1529,6 +1529,7 @@ void __libc_exit_fini()
 			fpaddr(p, dyn[DT_FINI])();
 #endif
 	}
+	__malloc_tls_teardown(self);
 }
 
 void __ldso_atfork(int who)
@@ -1866,6 +1867,9 @@ void __dls3(size_t *sp, size_t *auxv)
 	/* Activate error handler function */
 	error = error_impl;
 
+	/* Here we can initialize the allocator */
+	__malloc_init(__pthread_self());
+
 	/* If the main program was already loaded by the kernel,
 	 * AT_PHDR will point to some location other than the dynamic
 	 * linker's program headers. */
@@ -2085,9 +2089,12 @@ void __dls3(size_t *sp, size_t *auxv)
 	/* Actual copying to new TLS needs to happen after relocations,
 	 * since the TLS images might have contained relocated addresses. */
 	if (initial_tls != builtin_tls) {
-		if (__init_tp(__copy_tls(initial_tls)) < 0) {
+		void *mtls = __pthread_self()->malloc_tls;
+		pthread_t ns = __copy_tls(initial_tls);
+		if (__init_tp(ns) < 0) {
 			a_crash();
 		}
+		ns->malloc_tls = mtls;
 	} else {
 		size_t tmp_tls_size = libc.tls_size;
 		pthread_t self = __pthread_self();
diff --git a/mimalloc/readme.md b/mimalloc/readme.md
index 87cee98d..ed9e4d6a 100644
--- a/mimalloc/readme.md
+++ b/mimalloc/readme.md
@@ -938,3 +938,4 @@ free list encoding](https://github.com/microsoft/mimalloc/blob/783e3377f79ee82af
 * 2019-10-07, `v1.1.0`: stable release 1.1.
 * 2019-09-01, `v1.0.8`: pre-release 8: more robust windows dynamic overriding, initial huge page support.
 * 2019-08-10, `v1.0.6`: pre-release 6: various performance improvements.
+
diff --git a/src/env/__init_tls.c b/src/env/__init_tls.c
index 11a897ab..5e9962da 100644
--- a/src/env/__init_tls.c
+++ b/src/env/__init_tls.c
@@ -28,6 +28,7 @@ int __init_tp(void *p)
 	td->robust_list.head = &td->robust_list.head;
 	td->sysinfo = __sysinfo;
 	td->next = td->prev = td;
+	td->malloc_tls = __malloc_tls_default;
 	return 0;
 }
 
@@ -119,6 +120,7 @@ static void static_init_tls(size_t *aux)
 	Phdr *phdr, *tls_phdr=0;
 	size_t base = 0;
 	void *mem;
+	pthread_t self;
 
 	for (p=(void *)aux[AT_PHDR],n=aux[AT_PHNUM]; n; n--,p+=aux[AT_PHENT]) {
 		phdr = (void *)p;
@@ -179,10 +181,14 @@ static void static_init_tls(size_t *aux)
 	}
 
 	/* Failure to initialize thread pointer is always fatal. */
-	if (__init_tp(__copy_tls(mem)) < 0)
+	self = __copy_tls(mem);
+	if (__init_tp(self) < 0)
 		a_crash();
 
 	init_ifunc();
+
+	/* Initialize malloc. */
+	__malloc_init(self);
 }
 
 weak_alias(static_init_tls, __init_tls);
diff --git a/src/exit/exit.c b/src/exit/exit.c
index a6869b37..6a0b78fc 100644
--- a/src/exit/exit.c
+++ b/src/exit/exit.c
@@ -1,6 +1,7 @@
 #include <stdlib.h>
 #include <stdint.h>
 #include "libc.h"
+#include "pthread_impl.h"
 
 static void dummy()
 {
@@ -20,6 +21,7 @@ static void libc_exit_fini(void)
 	for (; a>(uintptr_t)&__fini_array_start; a-=sizeof(void(*)()))
 		(*(void (**)())(a-sizeof(void(*)())))();
 	_fini();
+	__malloc_tls_teardown(__pthread_self());
 }
 
 weak_alias(libc_exit_fini, __libc_exit_fini);
diff --git a/src/internal/pthread_impl.h b/src/internal/pthread_impl.h
index de2b9d8b..f8cf1e64 100644
--- a/src/internal/pthread_impl.h
+++ b/src/internal/pthread_impl.h
@@ -6,10 +6,15 @@
 #include <errno.h>
 #include <limits.h>
 #include <sys/mman.h>
+#ifndef MI_LIBC_BUILD
 #include "libc.h"
 #include "syscall.h"
 #include "atomic.h"
 #include "futex.h"
+#else
+/* restricted version for MI_LIBC_BUILD; need struct pthread + __pthread_self */
+#include <stdint.h>
+#endif
 
 #include "pthread_arch.h"
 
@@ -58,6 +63,7 @@ struct pthread {
 	volatile int killlock[1];
 	char *dlerror_buf;
 	void *stdio_locks;
+	void *malloc_tls;
 
 	/* Part 3 -- the positions of these fields relative to
 	 * the end of the structure is external and internal ABI. */
@@ -122,6 +128,7 @@ enum {
 #define __pthread_self() ((pthread_t)__get_tp())
 #endif
 
+#ifndef MI_LIBC_BUILD
 #ifndef tls_mod_off_t
 #define tls_mod_off_t size_t
 #endif
@@ -187,6 +194,11 @@ hidden void __tl_lock(void);
 hidden void __tl_unlock(void);
 hidden void __tl_sync(pthread_t);
 
+extern hidden void * const __malloc_tls_default;
+
+hidden void __malloc_init(pthread_t);
+hidden void __malloc_tls_teardown(pthread_t);
+
 extern hidden volatile int __thread_list_lock;
 
 extern hidden volatile int __abort_lock[1];
@@ -201,5 +213,6 @@ extern hidden unsigned __default_guardsize;
 #define DEFAULT_GUARD_MAX (1<<20)
 
 #define __ATTRP_C11_THREAD ((void*)(uintptr_t)-1)
+#endif
 
 #endif
diff --git a/src/malloc/calloc.c b/src/malloc/calloc.c
index bf6bddca..6aa482c6 100644
--- a/src/malloc/calloc.c
+++ b/src/malloc/calloc.c
@@ -32,6 +32,10 @@ weak_alias(allzerop, __malloc_allzerop);
 
 void *calloc(size_t m, size_t n)
 {
+#ifdef LIBC_CALLOC_EXTERNAL
+	if (!__malloc_replaced)
+		return __libc_calloc(m, n);
+#endif
 	if (n && m > (size_t)-1/n) {
 		errno = ENOMEM;
 		return 0;
diff --git a/src/malloc/external/empty.h b/src/malloc/external/empty.h
new file mode 100644
index 00000000..40a8c178
--- /dev/null
+++ b/src/malloc/external/empty.h
@@ -0,0 +1 @@
+/* empty */
diff --git a/src/malloc/libc_calloc.c b/src/malloc/libc_calloc.c
index d25eabea..3895c8cf 100644
--- a/src/malloc/libc_calloc.c
+++ b/src/malloc/libc_calloc.c
@@ -1,4 +1,8 @@
+#ifndef LIBC_CALLOC_EXTERNAL
+
 #define calloc __libc_calloc
 #define malloc __libc_malloc
 
 #include "calloc.c"
+
+#endif
diff --git a/src/malloc/mallocng/malloc.c b/src/malloc/mallocng/malloc.c
index ddad83ac..b8e38c74 100644
--- a/src/malloc/mallocng/malloc.c
+++ b/src/malloc/mallocng/malloc.c
@@ -7,6 +7,8 @@
 
 #include "meta.h"
 
+hidden void * const __malloc_tls_default = NULL;
+
 LOCK_OBJ_DEF;
 
 const uint16_t size_classes[] = {
diff --git a/src/thread/pthread_create.c b/src/thread/pthread_create.c
index 087f6206..1810e016 100644
--- a/src/thread/pthread_create.c
+++ b/src/thread/pthread_create.c
@@ -68,6 +68,7 @@ _Noreturn void __pthread_exit(void *result)
 	}
 
 	__pthread_tsd_run_dtors();
+	__malloc_tls_teardown(self);
 
 	__block_app_sigs(&set);
 
@@ -319,6 +320,7 @@ int __pthread_create(pthread_t *restrict res, const pthread_attr_t *restrict att
 	new->self = new;
 	new->tsd = (void *)tsd;
 	new->locale = &libc.global_locale;
+	new->malloc_tls = __malloc_tls_default;
 	if (attr._a_detach) {
 		new->detach_state = DT_DETACHED;
 	} else {
@@ -395,3 +397,8 @@ fail:
 
 weak_alias(__pthread_exit, pthread_exit);
 weak_alias(__pthread_create, pthread_create);
+
+static void mdummy(pthread_t p) {}
+
+weak_alias(mdummy, __malloc_init);
+weak_alias(mdummy, __malloc_tls_teardown);
-- 
2.25.1

